File: 502.ipo.cs
/*
 * @lc app=leetcode id=502 lang=csharp
 *
 * [502] IPO
 */

// @lc code=start
public class Solution 
{
    public int FindMaximizedCapital(int k, int w, int[] profits, int[] capital) 
    {
        List<int[]> couples = new List<int[]>();
        PriorityQueue<int, int> queue = new PriorityQueue<int, int>();
        int n = profits.Length;
        for (int j = 0; j < n; j++)
        {
            couples.Add(new int[2] {capital[j], profits[j]});
        }
        couples = couples.OrderBy(x => x[0]).ToList();
        int i = 0;
        while (k > 0)
        {
            while (i < couples.Count && couples[i][0] <= w)
            {
                queue.Enqueue(couples[i][1], - couples[i][1]);
                i++;
            }
            if (queue.Count == 0)
            {
                return w;
            }
            w += queue.Dequeue();
            k--;
        }
        return w;
    }
}
// @lc code=end

File: 502.ipo.py
#
# @lc app=leetcode id=502 lang=python3
#
# [502] IPO
#

# @lc code=start
import heapq
from typing import List


class Solution:
    '''
    First put the pairs together
    Sort according to capital
    if nothing in the heap, can stop here
    pop out the one with max profit
    '''
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        couples = []
        for i in range(len(capital)): couples.append((capital[i], profits[i]))
        couples.sort()
        heap = []
        i = 0
        while k > 0:
            while i < len(couples) and couples[i][0] <= w:
                heapq.heappush(heap, - couples[i][1])
                i += 1
            if not heap: return w
            w -= heapq.heappop(heap)
            k -= 1
        return w
                
                
# @lc code=end

